File: src/client-handlers/android.js
"use strict";
var util = require('util')
var path = require('path')

var pageDir = path.join(__dirname, '..', '..', 'pages')


var _BaseHandler = function _BaseHandler() {}

_BaseHandler.prototype.isConnectivityCheck = function(req) {
  return req.url === '/generate_204'
}

_BaseHandler.prototype.recognizes = function(req) {
  return req.get('User-Agent') && req.get('User-Agent').search('Android|android') !== -1
}


// Android client that opens all pages in CNA
var AndroidCnaHandler = exports.AndroidCnaHandler = function AndroidCnaHandler() { 
  _BaseHandler.apply(this, arguments) 
}
util.inherits(AndroidCnaHandler, _BaseHandler)

AndroidCnaHandler.prototype.run = function(client, req, res, next) { next() }


// Android client only dislay a message in CNA and a button to close itself
var AndroidBrowserHandler = exports.AndroidBrowserHandler = function AndroidBrowserHandler(options) { 
  _BaseHandler.apply(this, arguments)
  this.options = options || {}
  this.options.connectedPagePath = this.options.connectedPagePath 
    || path.join(pageDir, 'android', 'connected.html')
  this.fakeConnectivity = false
}
util.inherits(AndroidBrowserHandler, _BaseHandler)

AndroidBrowserHandler.prototype.run = function(client, req, res, next) {
  client.status = client.status || 'not-connected'

  // requests testing the connectivity of the network :
  // - when first connecting to the network, any other HTTP response than 204 will trigger the CNA to open
  //   and display the responded page.
  // - when CNA is open if HTTP 204 is sent the CNA will close
  if (this.isConnectivityCheck(req)) {
    if (client.status === 'not-connected')
      res.sendFile(this.options.connectedPagePath)
    else
      res.status(204).end()
  
  // On "connected.html", when the user clicks on the link, subsequent connectivity 
  // checks will be answered with 204, which will close the CNA.
  } else if (client.status === 'not-connected') {
    if (req.path === '/node-captive-portal/connected') {
      client.status = 'connected'
      res.end()
    } else next()
  
  } else next()
}

File: freedomportal/handlers/android.lua
local config = require('freedomportal.config')

local function recognizes(wsapi_env)
    return wsapi_env.HTTP_USER_AGENT and wsapi_env.HTTP_USER_AGENT:match('Android')
end

-- Android client that opens all pages in CNA
local function run_cna(client_infos, wsapi_env)
    return { code = 'PASS' }
end

-- Android client only dislay a message in CNA and a button to close itself
local function run_browser(client_infos, wsapi_env)

    -- requests testing the connectivity of the network :
    -- * when first connecting to the network, any other HTTP response than 204 will trigger 
    --      the CNA to open and display the responded page.
    -- * when CNA is open if HTTP 204 is sent the CNA will close
    if wsapi_env.PATH_INFO == '/generate_204' then
        if client_infos.status == nil then
            local location = config.get('captive_static_root_url') .. '/android/connected.html'
            return { code = 302, headers = { Location = location } }
        else
            return { code = 204 }
        end
      
    -- On "connected.html", when the user clicks on the link, subsequent connectivity 
    -- checks will be answered with 204, which will close the CNA.
    elseif client_infos.status == nil then
        if wsapi_env.PATH_INFO == config.get('captive_dynamic_root_url') .. '/connected' then
            return { code = 200, client_infos = { status = 'connected' } }
        else -- TODO : Does this ever happen?
            return { code = 200 }
        end

    else 
        return { code = 'PASS' }
    end
end

return {
    cna = { run = run_cna, recognizes = recognizes },
    browser = { run = run_browser, recognizes = recognizes },
}